File: examples/interpolation_demo.cxx
#include "concentric_interpolation.h"

/** \brief Concentric Interpolation demo
 *
 * This is an implementation of the example from Section 7.2 of the original
 * article [https://doi.org/10.1007/s10444-019-09726-5]
 *
 * Optimizes the kernel parameter on a subset of the validation data, using
 * relative errors on a spherical subset of the validation data.
 *
 * Then, computes the absolute error on the whole validation data.
 *
 */

int main()
{
    // set up support data
    char FilenameSupportDirections[]    = "../data/directions/D8_N166_from_N83_sym1_s-2.txt";
    char FilenameSupportRadii[]         = "../data/radii/radii_4.txt";
    char FilenameSupportValues[]        = "../data/values/D8_N166from83_s-2_R4.txt";
    ConcentricData SupportData( FilenameSupportDirections,
                                FilenameSupportRadii,
                                FilenameSupportValues );

    // set up the Concentric Interpolation scheme
    ConcentricInterpolation ConcentricInterpolant(SupportData);

    // set up validation data
    char FilenameValidationDirections[] = "../data/directions/D8_N512_from_N256_sym1_s-2.txt";
    char FilenameValidationRadii[]      = "../data/radii/radii_4.txt";
    char FilenameValidationValues[]     = "../data/values/D8_N512from256_s-2_R4.txt";
    ConcentricData ValidationData( FilenameValidationDirections,
                                   FilenameValidationRadii,
                                   FilenameValidationValues );


    // find the optimal gamma value for this validation set
    const double gamma_min = 0.1;
    const double gamma_max = 2.0;
    const int   N_gamma_regular   = 10;
    const int   N_gamma_bisection = 5;
    ConcentricInterpolant.OptimizeGamma(    gamma_min,
                                            gamma_max,
                                            N_gamma_regular,
                                            N_gamma_bisection,
                                            ValidationData,
                                            1,                  // use relative error, see ConcentricInterpolation::Error()
                                            0,                  // first component of values
                                            0,                  // last component of values
                                            2                   // only compute error on radius with this index (zero-based), see ConcentricInterpolation::Error()
                                       );

    // compute the error on the validation set, using the above found gamma
    printf("\nRMS absolute error on whole validation data = %e\n\n",
           ConcentricInterpolant.Error(ValidationData));

    return 0;
}

File: python/interpolation_demo.py
import sys
sys.path.append('../build/release/python')
from pyconinter import ConcentricData, ConcentricInterpolation

data_root = "../data/"
filename_support_directions = data_root + "directions/D8_N166_from_N83_sym1_s-2.txt"
filename_support_radii = data_root + "radii/radii_4.txt"
filename_support_values = data_root + "values/D8_N166from83_s-2_R4.txt"

support_data = ConcentricData(
    filename_support_directions, filename_support_radii, filename_support_values
)

concentric_interpolant = ConcentricInterpolation(support_data)
#concentric_interpolant.print_info()
print(concentric_interpolant.get_info())

filename_validation_directions = data_root + "directions/D8_N512_from_N256_sym1_s-2.txt"
filename_validation_radii = data_root + "radii/radii_4.txt"
filename_validation_values = data_root + "values/D8_N512from256_s-2_R4.txt"

validation_data = ConcentricData(
    filename_validation_directions,
    filename_validation_radii,
    filename_validation_values,
)

gamma_min = 0.1
gamma_max = 2.0
N_gamma_regular = 10
N_gamma_bisection = 5
concentric_interpolant.optimize_gamma(
    gamma_min,
    gamma_max,
    N_gamma_regular,
    N_gamma_bisection,
    validation_data,
    1,
    0,
    0,
    2,
)

print(f"\nRMS absolute error on whole validation data = {concentric_interpolant.compute_error(validation_data):e}\n\n")